provider/provider.go

```diff
@@ -213,25 +232,75 @@ func New(opts ...Option) (*SweepingProvider, error) {
 		connectivity.WithOfflineCheckInterval(cfg.connectivityCheckOfflineInterval),
 	)
 	if err != nil {
+		cleanup(prov.cleanupFuncs)
 		return nil, err
 	}
+	prov.cleanupFuncs = append(prov.cleanupFuncs, prov.connectivity.Close)
 
 	go prov.run()
 
 	return prov, nil
 }
 
+const retryInterval = 5 * time.Minute
+
 func (s *SweepingProvider) run() {
+	s.wg.Add(1)
```

Comment:
It is better to call `wg.Add` before the goroutine is started, not in the goroutine itself. Calling before the goroutine is started generally ensures that `wg.Wait` is not called before `wg.Add` due to goroutine scheduling. My not be strictly necessary here, but good coding practice.